WordPress/WordPressTest/SiteCreation/SiteCreationNameTracksEventTests.swift

```diff
@@ -0,0 +1,94 @@
+import XCTest
+import AutomatticTracks
+@testable import WordPress
+
+class SiteCreationNameTracksEventTests: XCTestCase {
+
+    private let siteNameEventPropertyKey = "site_name"
+
+    override func setUpWithError() throws {
+        TestAnalyticsTracker.setup()
+    }
+
+    override func tearDownWithError() throws {
+        TestAnalyticsTracker.tearDown()
+    }
+
+    func siteNameViewControllerMaker() throws -> SiteNameViewController {
+        let mockSiteCreator = SiteCreator()
+        let siteNameStep = SiteNameStep(creator: mockSiteCreator)
+        let siteIntentViewController = try XCTUnwrap(siteNameStep.content as? SiteNameViewController)
+        return siteIntentViewController
+    }
+
+    func load(_ siteNameViewController: SiteNameViewController) {
+        siteNameViewController.loadViewIfNeeded()
+        siteNameViewController.viewDidLoad()
+    }
+
+    func tap(_ barButtonItem: UIBarButtonItem?) throws {
+        let action = try XCTUnwrap(barButtonItem?.action)
+        UIApplication.shared.sendAction(action, to: barButtonItem?.target, from: nil, for: nil)
+    }
+
+    func testSiteNameTracksEventFiresWhenViewed() throws {
+
+        // Given
+        let siteNameViewController = try siteNameViewControllerMaker()
+        let expectedEvent = WPAnalyticsEvent.enhancedSiteCreationSiteNameViewed.value
+
+        // When
+        load(siteNameViewController)
+
+        // Then
+        let lastTrackedEvent = try XCTUnwrap(TestAnalyticsTracker.tracked.last?.event)
+        XCTAssertEqual(lastTrackedEvent, expectedEvent)
+    }
+
+    func testSiteNameTracksEventFiresWhenSkipped() throws {
+
+        // Given
+        let siteNameViewController = try siteNameViewControllerMaker()
+        let expectedEvent = WPAnalyticsEvent.enhancedSiteCreationSiteNameSkipped.value
+
+        // When
+        load(siteNameViewController)
+        try tap(siteNameViewController.navigationItem.rightBarButtonItem)
+
+        // Then
+        let trackedEvent = try XCTUnwrap(TestAnalyticsTracker.tracked.last?.event)
+        XCTAssertEqual(expectedEvent, trackedEvent)
+    }
+
+    func testSiteNameTracksEventFiresWhenCanceled() throws {
+
+        // Given
+        let siteNameViewController = try siteNameViewControllerMaker()
+        let expectedEvent = WPAnalyticsEvent.enhancedSiteCreationSiteNameCanceled.value
+
+        // When
+        siteNameViewController.viewMovingFromParent()
```

Comment:
It feels that in this test the `canceled` event is not exactly tested, because we are not firing the event (`if isMovingFromParent`), that should trigger the tracking, but we are sort of directly calling `TestAnalyticsTracker.trackSiteNameCanceled()`. The issue is that firing UI events should not really be part of unit tests, but on the other hand our UI tests, at least at the moment, are not equipped for this.

Reply:
I agree. When implementing this my mental framework was:
- At some point we have to trust the view code will do what it's supposed to do (e.g. that `viewMovingFromParent()` will get called.
- These UIKit classes have lots of public methods that no one should call
- Everything would be better if we could avoid UI code altogether, but I rather have _some_ tests than none, or undergo a massive refactor at this time.

Reply:
>It feels that in this test the canceled event is not exactly tested, because we are not firing the event (if isMovingFromParent), that should trigger the tracking

I wouldn't see this as a big blocker because I'm pretty used to doing this pattern in Android, from the times where MVP (Model View Presenter) was the popular pattern.

To sum up how automated testing would look like for MVP in this case:
- Ideally we'd have UI tests for the UI
- The unit tests for the presenter would test by calling directly the method that's supposedly called by the platform specific code (which is the view implementation class).

The trick to get that right is to move all the logic from the platform-specific code to the testable bits. In case something relies on things decided by the mobile OS, the only logic acceptable in the view would be the one taking care of mapping data from the view to a format accepted by the viewmodel/viewcontroller/presenter/(whatever other handler specific to the design pattern).